File: stralnumcmp.py
import re

stralnumcmp_re = re.compile(r'((\d+)|(\D+))')

def stralnumcmp(a0, b0):
    a = a0
    b = b0

    while len(a) > 0 and len(b) > 0:
        m_a = stralnumcmp_re.match(a)
        m_b = stralnumcmp_re.match(b)
        if m_a.group(2) and m_b.group(2):
            c = int(m_a.group(2)) - int(m_b.group(2))
        else:
            a1 = m_a.group(2) or m_a.group(3)
            b1 = m_b.group(2) or m_b.group(3)
            c = (a1 > b1) - (a1 < b1)
        if c != 0:
            return c
        a = a[m_a.end(1):]
        b = b[m_b.end(1):]

    return len(a0) - len(b0)

File: stralnumcmp.lua
function stralnumcmp(a, b)
    local a0, b0, an, bn, as, bs, c
    a0 = a
    b0 = b
    while (a:len() > 0 and b:len() > 0) do
	an = a:match('^%d+')
	bn = b:match('^%d+')
	as = an or a:match('^%D+')
	bs = bn or b:match('^%D+')

	if (an and bn) then
	    c = tonumber(an) - tonumber(bn)
	else
	    c = (as < bs) and -1 or ((as > bs) and 1 or 0)
	end
	if (c ~= 0) then
	    return c
	end
	a = a:sub((an and an:len() or as:len()) + 1)
	b = b:sub((bn and bn:len() or bs:len()) + 1)
    end
    return (a0:len() - b0:len())
end
